src/openms/include/OpenMS/CHEMISTRY/SimpleTSGXLMS.h

```diff
@@ -0,0 +1,309 @@
+       * @param link_pos_2 A second position for the linker, in case it is a loop link
+       */
+      virtual void addXLinkIonPeaks_(std::vector< SimplePeak >& spectrum, AASequence & peptide, Size link_pos, double precursor_mass, Residue::ResidueType res_type, std::vector< LossIndex > & forward_losses, std::vector< LossIndex > & backward_losses, int charge, Size link_pos_2 = 0) const;
+
+      /**
+       * @brief Adds cross-linked ions of a specific ion type and charge to a spectrum
+
+        This version of the function is for cross-linked peptide pairs.
+
+       * @param spectrum The spectrum to which the new peaks are added. Does not have to be empty, the generated peaks will be pushed onto it.
+       * @param crosslink The ProteinProteinCrossLink to be fragmented
+       * @param link_pos The position of the cross-linker on the given peptide
+       * @param precursor_mass The mass of the whole cross-link candidate or the precursor mass of the experimental MS2 spectrum.
+       * @param frag_alpha True, if the fragmented peptide is the Alpha peptide. Used for ion-name annotation.
+       * @param res_type The ion type of the added peaks
+       * @param forward_losses  vector of LossIndex generated by getForwardLosses_ for the fragmented peptide
+       * @param backward_losses vector of LossIndex generated by getBackwardLosses_ for the fragmented peptide
+       * @param losses_peptide2 set of losses for the second, not fragmented peptide, e.g. last set from getForwardLosses_ for the second peptide
+       * @param charge The charge of the added peaks
+       */
+      virtual void addXLinkIonPeaks_(std::vector< SimplePeak >& spectrum, OPXLDataStructs::ProteinProteinCrossLink & crosslink, bool frag_alpha, Residue::ResidueType res_type, std::vector< LossIndex > & forward_losses, std::vector< LossIndex > & backward_losses, LossIndex & losses_peptide2, int charge) const;
+
+      /**
+       * @brief Calculates sets of possible neutral losses for each position in the given peptide
+
+        This function generates a vector of sets. Each set contains the possible neutral losses for a specific prefix of the peptide.
+
+       * @param peptide The peptide or ion for which to collect possible losses
+       */
+      std::vector< LossIndex > getForwardLosses_(AASequence & peptide) const;
+
+      /**
+       * @brief Calculates sets of possible neutral losses for each position in the given peptide
+
+        This function generates a vector of sets. Each set contains the possible neutral losses for a specific suffix of the peptide.
+
+       * @param peptide The peptide or ion for which to collect possible losses
+       */
+      std::vector< LossIndex > getBackwardLosses_(AASequence & peptide) const;
+
+      // TODO copied from normal TSG, but it is protected over there. Move it to Residue class maybe?
+      /// helper for mapping residue type to letter
+      char residueTypeToIonLetter_(Residue::ResidueType res_type) const;
```

Comment:
hmm sounds as if it could be moved to Residue class

Reply:
Yes, I can do that. And adapt all spectrum generators to use that.